### ISSUE TITLE
Windows binaries with Unicode paths

### ISSUE THREAD

**Issue by ncruces** [Exiv2/exiv2#679]
Would it be too much to ask if Windows binaries of 0.27 were compiled with `EXV_UNICODE_PATH` enabled?

Does it somehow break expectations for library users?

It's a bit of a pain that none of the `exiv2.exe` binaries (msvc, mingw, cygwin) seem to support Unicode filenames out of the box.

**Comment by clanmills** [comment]
Good Question!  The only sample program that understands EXIV2_UNICODE_PATH is samples/exifprint.cpp.

About 5 years ago, I investigated compiling support for both wchar_t and char in a single build and concluded that this would be very difficult.  In my working life (at Adobe), I found writing code for UNICODE by default worked well, and getting ASCII to "widen" was straightforward.   However, (although I've forgotten the details), I concluded that Exiv2's UNICODE_PATH made "dual support" impossible.  

If you know how this can be achieved, I'm very interested to hear your ideas.

**Comment by ncruces** [comment]
Welp. I see this is more complicated than I imagined.

I'm using exiv2, the command-line utility, and I'm frustrated that none of the official Windows binaries seem to handle files/folders with Unicode characters in them.

I have to use the workaround of 8.3 short names in my scripts (where available), and give up otherwise.

I've looked around, found `EXV_UNICODE_PATH`, and it seemed to be the answer. 


To be honest, haven't tried compiling exiv2 myself, the process seems a bit daunting (which is always the case on Windows, but even compared to other C++ projects, exiv2 seems scary).

I'm sorry I'm not being very helpful. If anyone has any guidance on getting an exiv2.exe that accepts Unicode paths on the command line, it would be much appreciated.

**Comment by D4N** [comment]
@ncruces I am afraid that there is no other short term solution than to compile exiv2 yourself. Enabling unicode paths on Windows is on my Todo list, but it's getting constantly deferred by other issues being more urgent (and me not having a Windows computer).

**Comment by ncruces** [comment]
OK, I'll try to do so then.

Feel free to close this, or leave it open as a feature request.

**Comment by D4N** [comment]
Let's keep this open as a reminder.

@ncruces In case you run into problems, documentation wise or others, please report them. It helps us a great deal.

**Comment by clanmills** [comment]
I believe we'll have to put in a considerable effort to get exiv2(.exe) to work in UNICODE.  When Windows executes command-line programs in wchar_t mode, it calls `wmain(int argc,wchar_t* argvw[])` so the command-lines have to be parsed in unicode.  There's a trick used in Windows concerning _t to enable code to be compiled for char or wchar_t.  I did this to samples/exifprint.cpp to test UNICODE_PATH for sanity.  However more work is required to "widen" exiv2.cpp and actions.cpp.

Why have I not done this?  There are always been higher priority matters demanding attention.  I don't believe this is difficult.  If you'd like to get involved with Exiv2, we'd be delighted to accept your contribution.

I'm interest in your comments about how difficult it is to build exiv2 on Windows.  You have to install python3, conan and cmake.  Then it's only three commands (one conan command and two cmake commands).  We used to provide Visual Studio .sln and .vcxproj files.  With v0.27, we decided to only support CMake and added support for conan to build dependencies (usually expat and zlib).  It's always challenging to use unfamiliar tools, however conan and cmake work well and work on all our supported platforms.   A little daunting.  Worth the effort.

**Comment by tbeu** [comment]
I actually wonder why we have two alternative interfaces? The exiv2.exe on Win can always be compiled with `EXV_UNICODE_PATH`. And the exiv2.dll can provide both interface functions in parallel instead of alternative.

**Comment by clanmills** [comment]
@tbeu I don't know why it was implemented as alternative, instead of parallel.  About 5 years ago, I did try to set up parallel and decided that this was quite a change.  However, I only spent about 2 or 3 hours before giving up.  Perhaps in v0.28, all Windows (msvc and MinGW) builds should offer APIs for both ascii and wide.

Perhaps we should think about a 5 minute video on YouTube to demonstrate installing and using CMake and Conan to build Exiv2.  I had email today with Johnathan Sachs (author of Lotus-1-2-3) and he was advocating for Visual Studio sln/vcxproj files.  If we make CMake/Conan look easy, we could reduce the fear that Windows developers have of Unix tools.

We're on vacation in Florida.  4 weeks of sunshine.  So happy to see you working with Michal, Dan, Luis and others.  Thanks for contributing.

**Comment by axxel** [comment]
I've just spend half a week updating my project (which is a rather convoluted bunch of Qt-Gui + plain cli apps + a bunch of 'frozen' python scripts) to support unicode filenames on windows and I learned two things that might be of interest to someone working on this:
 a) you get unicode (utf-8 based) support through your existing 'narrow' apis for free if you enable the relatively new check box "Beta:  Use Unicode UTF-8 for worldwide language support" in your windows 10 language settings.
 b) this development is nicely in line with what http://utf8everywhere.org/ is propagating.

Bottom line: don't go the narrow + wide route (in the sense like windows does it with the ...A vs. ...W 'overloads' in their API and sprinkling `_t...` macros all over your source code). Instead, either simply wait it out (see a)) or take the approach in b) and only deal with wide strings directly interfacing windows apis (see e.g. Boost::nowide).

**Comment by clanmills** [comment]
@axxel.  I agree we "wait it out".   I was mistaken about what the cmake option -DEXIV_UNICODE_PATH=On does.  It augments the API by adding UNICODE entry point to several function such as `Exiv2::ImageFactory::open()`

So the exiv2 samples don't need to go down the A/W road used in Win32.

```
        static Image::AutoPtr open(const std::string& path, bool useCurl = true);
#ifdef EXV_UNICODE_PATH
        /*!
          @brief Like open() but accepts a unicode path in an std::wstring.
          @note This function is only available on Windows.
         */
        static Image::AutoPtr open(const std::wstring& wpath, bool useCurl = true);
#endif
```

It's an interesting idea that we should make that DEXIV_UNICODE_PATH=On the default for all windows builds.

I'm going to close this.  If somebody really wants this, the issue can be reopened.
